### PR TITLE
Port to windows

### PR DESCRIPTION
**Pull request by Malacath-92**
I ported your the tools to run on Windows. To make this work without too much of a hack I had to make a couple adjustments:
- Make the scripts folder into a python module, so that it doesn't have to be invoked with subprocess, but can just be imported. This was required so that the different executions share an environment, which in turn was required since compiling with MSVC works only when invoking vcvarsall.bat beforehand, which sets a bunch of environment variables. This bat file is too slow to feasibly execute on every compilation. On the other hand, this forces to compile only for one version of MSVC at a time, since there is no guarantee that different versions of msvc can compile correctly with the env of other versions.
- Renamed all scripts to use underscores instead of bashes, to allow importing them.
- Made sure that all the scripts can still be executed standalone, as before.
- Add the find_visual_studio.py file which finds one of MSVC 2015, 2017 or 2019 based on some registry variables.
- Deferred most "generation" of cmd line args to the compiler to analyze_file.py

Unfortunately these changes mean that the diff with master is useless. So I can either make a separate PR for the "modularization" of the scripts and after that make a PR for the windows port or you can look at the separate submits.

I was also unsure about what style I should follow, but I tried to adjust to the style from existing source.

Currently supports one of MSVC 2015, 2017 or 2019 (c++14 & c++17) as well as clang and clang-cl (c++11, c++14 & c++17)

[Example run (just for proof of concept that the PR works, since my laptop was unplugged for parts of the run)](https://github.com/Philip-Trettner/cpp-compile-overhead/files/4671878/compile-health-data.txt)

### CODE REVIEW

**Comment by Philip-Trettner** [comment]
First of all, thanks a lot!

I'll check if it works on my systems as well and merge both changes at once if it does. My expertise is more in C++ so sorry for the Python to be a bit messy ;)

File: scripts/generate_jobs.py [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+            self.args = []
+
+
+    def generate_configs():
+
+        msvc_variants = [
+            ['Debug', ['/Od', '/Ob0', '/MDd', '/GS', '/DWIN32', '/D_WINDOWS', '/DNDEBUG']],
+            ['RelWithDebInfo', ['/O2', '/Ob1', '/MD', '/GS', '/DWIN32', '/D_WINDOWS', '/DNDEBUG']],
+            ['Release', ['/O2', '/Ob2', '/MD', '/GS', '/DWIN32', '/D_WINDOWS', '/DNDEBUG']],
+        ]
```
Comment (Philip-Trettner) [inline scripts/generate_jobs.py]: is it correct that all variants define `NDEBUG`?

Comment (Malacath-92) [reply scripts/generate_jobs.py]: Definitely not  😅

Comment (Malacath-92) [reply scripts/generate_jobs.py]: I will have to check later when I get off work, but I believe only the Debug configuration does not define `NDEBUG`

**Comment by Malacath-92** [comment]
@Philip-Trettner
Is there anything that blocks this from going forward. I can be patient or fix any other issues that need immediate attention. I just thought I'd check in since it has been two weeks.

**Comment by Philip-Trettner** [comment]
Sorry for the delay, had a lot to do and this PR slipped my mind a bit.

I'm trying to run it on linux but `import winreg` is not working (and neither is `pip3 install winreg`). Is there something like a conditional `import scripts.find_visual_studio`?

**Comment by Malacath-92** [comment]
Yes, that's my bad. I'm not a python savant either but I believe we can use regular control flow (i.e. if statements) to conditionally import. If my commit just now doesn't work I will check after work.
To be fair, I should test this PR on Linux myself. Maybe I will do that I hand it to you when I had the chance to do that.
I'm sorry if I appeared impatient.

**Comment by Philip-Trettner** [comment]
It seems to work on linux now. It really is a shame that we cannot do more than one msvc version at the same time though.

**Comment by Malacath-92** [comment]
I have some ideas how to do more than one msvc version, but the only once that I can think of which don't execute vcvarsall.bat every compilation would be quite intrusive to job execution as it stands now.
Namely I would associate with each job a pre-job script that has to be executed (in this context that'd be a call to the associated vcvarsall.bat, depending on the version of MSVC) and then batch jobs by that pre-job script. We could then execute each batch in it's own, clean environment.
All linux based tests would then just not be associated to any pre-job script and thus be batched as one.

I'm not sure how much work this would be, but we could open an issue and discuss possible solutions to see if maybe we can come up with something else.

**Pull request merged**